### PR TITLE
Make fuse_all_reduce_op_pass support mix_precision

### PR DESCRIPTION
**Pull request by chengduoZH**
1. Polish fuse_all_reduce_op_pass and alloc_continuous_space_for_grad_pass to clean up  not used `SetGroupAccordingToGroupSize` function.
2. make fuse_all_reduce_op_pass support mixed dtype such as fp32 and fp16 all.

### CODE REVIEW

File: paddle/fluid/operators/ngraph/ops/activation_op.h [hunk 1]
```diff
@@ -113,6 +113,5 @@ void BuildTanhGradNode(
 REGISTER_NG_OP(gelu, BuildGeluNode);
 REGISTER_NG_OP(gelu_grad, BuildGeluGradNode);
 REGISTER_NG_OP(relu_grad, BuildReluGradNode);
-REGISTER_NG_OP(sqrt, BuildUnaryNode<ngraph::op::Sqrt>);
```
Comment (gongweibao) [inline paddle/fluid/operators/ngraph/ops/activation_op.h]: Why delete this?

File: paddle/fluid/framework/details/multi_devices_helper.h [hunk 2]
```diff
@@ -58,15 +58,15 @@ constexpr char kFusedVarNamePrefix[] = "@FUSEDVAR@";
...
+typedef std::vector<std::string> FusedGrads;
 constexpr char kFusedGrads[] = "fused_gradients";
 
 typedef std::vector<std::pair<std::string, std::string>> ParamsAndGrads;
 constexpr char kParamsAndGrads[] = "params_grads";
 
 typedef std::vector<std::vector<std::pair<std::string, std::string>>>
-    GroupGradsAndParams;
-constexpr char kGroupGradsAndParams[] = "group_grads_params";
+    GroupParamsAndGrads;
```
Comment (gongweibao) [inline paddle/fluid/framework/details/multi_devices_helper.h:68]: 👍

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc [hunk 3]
```diff
@@ -71,81 +70,98 @@ class AllocContinuousSpaceForGradPass : public ir::Pass {
...
-    for (auto &node : topo_nodes) {
-      RecordParamsAndGrads(node, &params_grads);
-    }
+    RecordParamsAndGrads(result, &params_grads);
 
-    if (params_grads.size() == 0) {
-      VLOG(10) << "Doesn't find gradients";
+    size_t num_params_grads = params_grads.size();
+    VLOG(10) << "The number of params and grads is:" << params_grads.size();
+    if (num_params_grads == 0) {
```
Comment (gongweibao) [inline paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc:81]: LOG(WARNING) here？

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc [hunk 4]
```diff
@@ -163,187 +192,227 @@ class AllocContinuousSpaceForGradPass : public ir::Pass {
...
       const details::ParamsAndGrads &params_grads,
-      details::GroupGradsAndParams *group_grads_params) const {
-    std::unordered_map<std::string, std::vector<int>> layer_params;
+      details::GroupParamsAndGrads *group_params_grads) const {
+    using var_dtype = std::pair<std::string, proto::VarType::Type>;
+    std::map<var_dtype, size_t> var_idx;
 
     for (size_t i = 0; i < params_grads.size(); ++i) {
       auto pos = params_grads[i].first.find_first_of(".");
+
```
Comment (gongweibao) [inline paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc:220]: It's not persistable if use memory optimization？

**Review by gongweibao — CHANGES_REQUESTED** [review]
We need a unit test to guarantee correctness.

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc (hunk 3 above)
Comment (chengduoZH) [reply paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc:81]: I think it doesn't matter.

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc (hunk 4 above)
Comment (chengduoZH) [reply paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc:220]: `params_grads[i].first` is the parameter's name, and it is persistable.

**Comment by chengduoZH** [comment]
>  We need a unit test to guarantee correctness.

@gongweibao Done

File: paddle/fluid/framework/ir/multi_devices_graph_pass/fuse_all_reduce_op_pass.cc [hunk 5]
```diff
@@ -103,6 +81,35 @@ class FuseAllReduceOpPass : public ir::Pass {
...
   }
 
+  std::unordered_map<std::string, Node *> GetAllReduceOps(
+      const Graph &result, const std::vector<platform::Place> &places,
+      const std::unordered_set<std::string> &grads) const {
+    size_t num_place = places.size();
+    std::unordered_map<std::string, Node *> all_reduce_ops;
+    all_reduce_ops.reserve(grads.size());
+    for (auto &node : result.Nodes()) {
+      if (node->IsOp()) {
```
Comment (gongweibao) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/fuse_all_reduce_op_pass.cc:91]:
```
if(!node->IsOp())    continue;
```

File: paddle/fluid/framework/ir/multi_devices_graph_pass/fuse_all_reduce_op_pass.cc [hunk 6]
```diff
@@ -103,6 +81,35 @@ class FuseAllReduceOpPass : public ir::Pass {
...
+        PADDLE_ENFORCE(node->IsWrappedBy<details::OpHandleBase>());
+        auto *all_reduce_op_handle = dynamic_cast<details::AllReduceOpHandle *>(
+            &node->Wrapper<details::OpHandleBase>());
+        if (all_reduce_op_handle) {
```
Comment (gongweibao) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/fuse_all_reduce_op_pass.cc:95]:
```
if(nullptr == all_reduce_op_handle) continue;
```

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc [hunk 7]
```diff
@@ -61,8 +62,9 @@ void SetFuseParameterMemorySize(double memory_size) {
 
 double GetFuseParameterMemorySize() { return FLAGS_fuse_parameter_memory_size; }
 
-static framework::proto::VarType::Type kDefaultDtype =
-    framework::proto::VarType::Type::VarType_Type_BOOL;
+// static const char kUnKnow[] = "@UNKNOW@";
+// static framework::proto::VarType::Type kDefaultDtype =
+// framework::proto::VarType::Type::VarType_Type_BOOL;
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc]: Please remove line65~66

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc [hunk 8]
```diff
@@ -175,57 +200,56 @@ class AllocContinuousSpaceForGradPass : public ir::Pass {
...
+  void SetGroupParamsAndGrads(
       const std::unordered_map<std::string, ir::Node *> &var_nodes,
       const details::ParamsAndGrads &params_grads,
-      details::GroupGradsAndParams *group_grads_params) const {
-    SetGroupAccordingToLayers(var_nodes, params_grads, group_grads_params);
-    SetGroupAccordingToMemorySize(var_nodes, group_grads_params);
+      details::GroupParamsAndGrads *group_params_grads) const {
+    SetGroupAccordingToLayers(var_nodes, params_grads, group_params_grads);
+    SetGroupAccordingToMemorySize(var_nodes, group_params_grads);
+    SetGroupAccordingToGroupSize(var_nodes, group_params_grads);
```
Comment (gongweibao) [inline paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc]: This function is not needed now.
```
SetGroupAccordingToGroupSize
```

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc [hunk 9]
```diff
@@ -249,88 +273,156 @@ class AllocContinuousSpaceForGradPass : public ir::Pass {
...
-                         group_grads_params->at(j).end());
+
+        group_p_g.insert(group_p_g.end(), group_params_grads->at(j).begin(),
+                         group_params_grads->at(j).end());
+
         ++j;
-        if (GetFuseParameterGroupsSize() > 1 &&
-            group_p_g.size() >
-                static_cast<size_t>(GetFuseParameterGroupsSize())) {
+        if (local_group_memory_size >= group_memory_size ||
```
Comment (gongweibao) [inline paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc]: Use left lines 281~290

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**